### ISSUE TITLE
Push messages remain in ApnsChannelPool

### ISSUE THREAD

**Issue by arjenvdhave** [jchambers/pushy#831]
We have this weird issue on our production system where at one moment push messages are not sent any more.
I tried to setup a test on my local machine which is running a mock server with SSL, and then I can reproduce the same behavior.

This is because the mock server is handling the SSL handshake ok, but after that it sents a GOAWAY instead of a SETTINGS frame.

This causes that the `GenericFutureListener` in the  `ApnsChannelPool::acquireWithinEventExecutor` is not called, and all push messages which are being sent end up in the `pendingAcquisitionPromises`

I think this can be fixed by overriding the ` public ChannelFuture goAway(final ChannelHandlerContext ctx, final int lastStreamId, final long errorCode, final ByteBuf debugData, ChannelPromise promise)` in the `ApnsClientHandler`

From which you can call the `getChannelReadyPromise(ctx.channel()).tryFailure`

This way the caller gets notified that the message was not sent.

I found out about this because of the following log line:
`2020-11-09 16:39:05,055 DEBUG [io.netty.handler.codec.http2.Http2ConnectionHandler] (nioEventLoopGroup-2-1) [id: 0x45cf0be4, L:/127.0.0.1:58473 - R:localhost/127.0.0.1:8081] Sent GOAWAY: lastStreamId '0', errorCode '1', debugData ''. Forcing shutdown of the connection.`

**Comment by jchambers** [comment]
Very strange—thank you for the report! Do you have code you can share for how you reproduced the issue with the mock server?

**Comment by arjenvdhave** [comment]
I used a simple http-server from nodejs, just to see if I could setup a http connection (https://www.npmjs.com/package/http-server). I used some generated certs and added this cert to the ClientBuilder with `setTrustedServerCertificateChain`

`http-server --ssl --cert ./localhost.crt --key ./localhost.key -p 8081`

This way I was able to pass the SSL handshake, but then the connection was closed.
This causes the infinite wait.

My java push code:

```
ApnsClient apnsClient = apnsClientBuilder.setClientCredentials(certificateStream, getCertKey())
				.setMetricsListener(applePushClientMetricsListener)
				.setConnectionTimeout(Duration.ofSeconds(10))
				.setFrameLogger(new Http2FrameLogger(LogLevel.TRACE))
				.setTrustedServerCertificateChain(getClass().getResourceAsStream("certificates/localhost.crt"))
				.build();
PushNotificationFuture<ApnsPushNotification, PushNotificationResponse<ApnsPushNotification>> sendFeature = apnsClient.sendNotification(pushNotification);

sendFeature.whenCompleteAsync((response, ex) -> {
	LOG.info("Response received");
}, executorService);
```

**Comment by HirumaStu** [comment]
Hi @jchambers, today I faced with similar problem. The reason is http-server doesn't support http/2.
Client received: io.netty.handler.codec.http2.Http2Exception: First received frame was not SETTINGS. Hex dump for first 5 bytes: 485454502f. It'd be useful to log some meaningful message on goaway event

**Comment by arjenvdhave** [comment]
Any idea's about this issue?
We just had the same issue on our server, the push messages just stopped after we had a firewall restart (due to a crash)

So my guess is that once  a connection is not returning the expected SETTING frame the client completly freezes.

**Comment by jchambers** [comment]
Please pardon the delay; it's been a very busy month. I hope to catch up on this (and your pull request!) shortly.

**Comment by arjenvdhave** [comment]
Just had another ddos on our server running our custom version with my PR
Sadly it didn't work

So when the lib isn't able to create a connection to the apple server it somehow starts putting all push messages in the queue but never recovers from this

Mayber there is a way to add some periodic check on the queue to see if it is growing and check if a new connection can be setup?

**Comment by jchambers** [comment]
I believe this should be fixed in #854.
